File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/gdpr/Gdpr.java
package com.snowplowanalytics.snowplow.internal.gdpr;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RestrictTo;

import com.snowplowanalytics.snowplow.internal.constants.TrackerConstants;
import com.snowplowanalytics.snowplow.payload.SelfDescribingJson;
import com.snowplowanalytics.snowplow.internal.utils.Preconditions;
import com.snowplowanalytics.snowplow.util.Basis;

import java.util.HashMap;
import java.util.Map;

@RestrictTo(RestrictTo.Scope.LIBRARY)
public class Gdpr {

    public final Basis basisForProcessing;
    public final String documentId;
    public final String documentVersion;
    public final String documentDescription;

    public Gdpr(@NonNull Basis basisForProcessing, @Nullable String documentId, @Nullable String documentVersion, @Nullable String documentDescription) {
        this.basisForProcessing = basisForProcessing;
        this.documentId = documentId;
        this.documentVersion = documentVersion;
        this.documentDescription = documentDescription;
    }

    @NonNull
    public SelfDescribingJson getContext() {
        Map<String, Object> map = new HashMap<>();
        map.put("basisForProcessing", basisForProcessing.toString().toLowerCase());
        map.put("documentId", documentId);
        map.put("documentVersion", documentVersion);
        map.put("documentDescription", documentDescription);
        return new SelfDescribingJson(TrackerConstants.SCHEMA_GDPR, map);
    }
}

File: snowplow-tracker/src/main/java/com/snowplowanalytics/core/gdpr/Gdpr.kt
/*
 * Copyright (c) 2015-2023 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */
package com.snowplowanalytics.core.gdpr

import androidx.annotation.RestrictTo
import com.snowplowanalytics.core.constants.TrackerConstants
import com.snowplowanalytics.snowplow.payload.SelfDescribingJson
import com.snowplowanalytics.snowplow.util.Basis

import java.util.*

@RestrictTo(RestrictTo.Scope.LIBRARY)
class Gdpr(
    val basisForProcessing: Basis,
    val documentId: String?,
    val documentVersion: String?,
    val documentDescription: String?
) {
    val context: SelfDescribingJson
        get() {
            val map: MutableMap<String, Any?> = HashMap()
            map["basisForProcessing"] =
                basisForProcessing.toString().lowercase(Locale.getDefault())
            map["documentId"] = documentId
            map["documentVersion"] = documentVersion
            map["documentDescription"] = documentDescription
            return SelfDescribingJson(TrackerConstants.SCHEMA_GDPR, map)
        }
}
